app/scripts/modules/core/cluster/cluster.service.ts

```diff
@@ -0,0 +1,236 @@
+            data[idx] = matches[0];
+          }
+        }
+      });
+
+      toRemove.forEach((idx) => data.splice(idx, 1));
+
+      // add any new ones
+      serverGroups.forEach((serverGroup) => {
+        if (!application.serverGroups.data.filter((test: ServerGroup) =>
+            test.name === serverGroup.name &&
+            test.account === serverGroup.account &&
+            test.region === serverGroup.region &&
+            test.category === serverGroup.category
+          ).length) {
+          data.push(serverGroup);
+        }
+      });
+      return data;
+    } else {
+      return serverGroups;
+    }
+  }
+
+  public createServerGroupClusters(serverGroups: ServerGroup[]): ICluster[] {
+    const clusters: ICluster[] = [];
+    const groupedByAccount = groupBy(serverGroups, 'account');
+    forOwn(groupedByAccount, (accountServerGroups, account) => {
+      const groupedByCategory = groupBy(accountServerGroups, 'category');
+      forOwn(groupedByCategory, (categoryServerGroups, category) => {
+        const groupedByCluster = groupBy(categoryServerGroups, 'cluster');
+        forOwn(groupedByCluster, (clusterServerGroups, clusterName) => {
+          const cluster: ICluster = {
+            account: account,
+            category: category,
+            name: clusterName,
+            serverGroups: clusterServerGroups,
+            cloudProvider: clusterServerGroups[0].cloudProvider
+          };
+          this.addHealthCountsToCluster(cluster);
+          clusters.push(cluster);
+        });
+      });
+    });
+    this.addProvidersAndServerGroupsToInstances(serverGroups);
+    return clusters;
+  }
+
+  public addExecutionsToServerGroups(application: Application): void {
+    let executions = get(application, 'runningExecutions.data', []);
+
+    if (!application.serverGroups.data) {
+      return; // still run if there are no running tasks, since they may have all finished and we need to clear them.
+    }
+
+    application.serverGroups.data.forEach((serverGroup: ServerGroup) => {
+      serverGroup.runningExecutions = [];
+      executions.forEach((execution: IExecution) => {
+        this.findStagesWithServerGroupInfo(execution.stages).forEach((stage: IExecutionStage) => {
+          const stageServerGroup = stage ? this.extractServerGroupNameFromContext(stage.context) : '';
+          const stageAccount = stage && stage.context ? stage.context.account || stage.context.credentials : '';
+          const stageRegion = stage ? this.extractRegionFromContext(stage.context) : '';
+
+          if (stageServerGroup.includes(serverGroup.name) &&
+            stageAccount === serverGroup.account &&
+            stageRegion === serverGroup.region) {
+              serverGroup.runningExecutions.push(execution);
+          }
+        });
+      });
+    });
+  }
+
+  public addTasksToServerGroups(application: Application): void {
+    let runningTasks = get(application, 'runningTasks.data', []);
```

Comment:
Should we just add the linter rule for this?

https://palantir.github.io/tslint/rules/prefer-const/